content/post/2020-02-07-processx-blocked-sigchld-fix/index.Rmd

```diff
@@ -0,0 +1,442 @@
+)
+```
+
+Still, it is possible (however unlikely :wink:) that the bug
+is in the parallel package, e.g. because it does not set up the signal
+handler for `SIGCHLD` properly, when the R session was started by
+processx/callr.
+
+The parallel package sets up a
+[finalizer, that runs when R exits](https://github.com/wch/r-source/blob/50dca8f210058532ee0837fad69b1ae78dcac23e/src/library/parallel/R/zzz.R#L33).
+This finalizer tries to eliminate all sub-processes that were started by
+parallel, and if it fails to do that, it emits the error message that we
+see above.
+
+When eliminating sub-processes, parallel sends them a `SIGKILL` signal, which
+is not possible to catch and handle, so it is sure that their execution
+has finished, and they are in a zombie state. In fact, `mclapply()` already
+tries to clean up the subprocesses it has started, so they are probably
+already in the zombie state when `mclapply()` returns. This is
+easy to check with the [ps](https://ps.r-lib.org/) package:
+
+```{r, cache = TRUE}
+tmp <- callr::r(
+  function() {
+    parallel::mclapply(1:2, function(x) x)
+    print(sapply(ps::ps_children(ps::ps_handle()), ps::ps_status))
+  },
+  show = TRUE
+)
+```
+
+Indeed, both sub-processes of the callr R process are zombies.
+Clearly, the callr R process did not receive or did not handle their
+`SIGCHLD` signals. To make sure that the `SIGCHLD` signal handler is
+properly set up in parallel, we need to debug parallel's C code, in the
+callr subprocess.
+
+We will use the [lldb](https://lldb.llvm.org/) debugger here, as that is
+the default on macOS. gdb has similar functionality, you probably want to
+use gdb for gcc and Linux.
+
+To debug the callr subprocess, we tell lldb to wait for a process called R:
+```
+❯ lldb -w -n R
+(lldb) process attach --name "R" --waitfor
+```
+
+Then in another terminal, we run `callr::r()`:
+```r
+callr::r(function() parallel::mclapply(1:2, function(x) x))
+```
+
+As soon as the callr subprocess starts, lldb will stop it (using a `SIGSTOP`
+signal that cannot be caught):
+
+```
+❯ lldb -w -n R
+(lldb) process attach --name "R" --waitfor
+Process 4196 stopped
+* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
+    frame #0: 0x00007fff74b75a4b libsystem_info.dylib`xdrmem_getlong_aligned + 75
+libsystem_info.dylib`xdrmem_getlong_aligned:
+->  0x7fff74b75a4b <+75>: callq  0x7fff74b759f0            ; _OSSwapInt32
+    0x7fff74b75a50 <+80>: movq   -0x18(%rbp), %rdx
+    0x7fff74b75a54 <+84>: movl   %eax, (%rdx)
+    0x7fff74b75a56 <+86>: movq   -0x10(%rbp), %rdx
+Target 0: (R) stopped.
+
+Executable module set to "/Library/Frameworks/R.framework/Resources/bin/exec/R".
+Architecture set to: x86_64h-apple-macosx-.
+(lldb)
+```
+
+Now we can set break points on
+[parallel C functions](https://github.com/wch/r-source/blob/50dca8f210058532ee0837fad69b1ae78dcac23e/src/library/parallel/src/fork.c).
+For a better debugging experience, it makes sense to recompile R without
```

Comment:
Would a link to https://github.com/wch/r-debug/blob/master/debugging-r.md make sense?

Reply:
I'll add some link at the end of the post.